### ISSUE TITLE
Corrupted index if reference genome starts with a "gap-only" scaffold

### ISSUE THREAD

**Issue by nigiord** [BenLangmead/bowtie2#192]
Problem
-----------
`bowtie2-build` silently creates a corrupted index if the first reference sequence in the provided fasta contains only gaps (series of Ns). Trying to map reads on such an index using `bowtie2` results in the following error:
``` bash
(ERR): bowtie2-align died with signal 11 (SEGV)
```

How to reproduce
-----------------------
Save the following genome into a file named testseq.fna.
``` fasta
>DUMMY_0
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNN
>DUMMY_1
AAAGACTGCGCAATTTCTGAAAACTGAATTCGAATAAATTTAGGGTTGAAGTCATTTAAA
TTTCCTGTGAAGTTGTTGGTTCGTGTGTTGATAAGTGATGAGCTGATTGACTGATACGAA
TTGTTTTGACGCTTGCATCTTTAGACTTTTTTGTACGGGGTGTTGCAAGCCAGTTCTCAA
ACAATGCGTGCCAAGATCCAGCCAGTCAAAGCGATGTGACTCAGCGTCACGAGCACAGAT
GCACCGTGCAGCCATTTGAAACGCAGCCGGTCATCAGAGTCAGTGGCTCGATTGATTGCT
GGCATCAGCAACTGCCGCACGGGGATGGTCGTTAAAGCAACGGCCGCCATAACAATTGCA
GCCATGGTGTCGTGAGACCACCACAATAAAGCGGCAACGGCCGCCGTAGCCACCACAAAC
AAATAGAACCAAGGAAAAGCAAGCCGCAGCGTAGCACCAGCAGATGGGGCTGGAAGTGCT
TTGAATAAAAAAGCAGCAAAGCCAAAGGAGTACAGGGTCATGCCGCCGAACAACAGCGCG
GTGACCAATAT
>DUMMY_2
TGGATAGGCAAAAAACAAAAAAGCCGCCTGGTGAACCAGGCGGCTTTGTTGGGGTGAATT
GACAGAACTTTTGTTCAAAGGCATAAAAACATTATGCACCAGCAACGTGACACCCTCGAA
CAGCGGTATCAAGCAGAAGTAGTGGCCTTTTTCAAATCCTTGACGGCTTGAAGAACTTCC
TCAACATGGCCTGGGACTTTGAGGCCTCTCCATTCGTTTTTGAGGGTGCCATCAGGGCCG
ACCAAAAATGTACTGCGCTCAATGCCTTTGACTTTCTTGCCGTACATGATTTTGTTTTTG
ACCACGCCAAACATGTGGCACATCTTTTCTTCCGTGTCCGCAATGAGTTCAAAGGGCAAC
TCAAGTTTGGCTTTGAAATCATCGTGGGACTTCATGTTATCGCGTGAGACACCCAAAACA
AGAGCACCAGCCTTGACGAAATCTTTGTATTTGTCGCGGAACTGCATGGCTTCTGTGGTG
CAACCCGGTGTGTTGTCTTTCGGGTAGAAAAACAAAACAACGGTTTGGCCGACGTGAGAG
ATGTTCGAAACTTTCAATCCGCTGGTGGCGTTGGCTTCAAATTCGGGGAGGGGTTTGTTC
AGAACGATCGCCATGGAACTTTAACTCGCTAGTCTGGTTGTGCGTACAGCACCTGCATGT
GGTGCCATTGCCGCTGGAGGCCAATAACCTCGGCAACCAGGTATTTTACTCTGAACGGCG
TTTTGACAGGCAACAACTGCCCA
```
Then
``` bash
bowtie2-build testseq.fna test-corrupted
```
The index is created without error. When trying to inspect the created index, one obtains the following output:
``` bash
$ bowtie2-inspect -v -n test-corrupted 
Input ebwt file: "test-corrupted"
Output file: ""
Local endianness: little
Assertions: enabled
Trying test-corrupted
  About to open input files: 11:09:13
Opening "test-corrupted.1.bt2"
Opening "test-corrupted.2.bt2"
  Finished opening input files: 11:09:13
  Reading header: 11:09:13
Headers:
    len: 1294
    bwtLen: 1295
    sz: 324
    bwtSz: 324
    lineRate: 6
    offRate: 4
    offMask: 0xfffffff0
    ftabChars: 10
    eftabLen: 20
    eftabSz: 80
    ftabLen: 1048577
    ftabSz: 4194308
    offsLen: 81
    offsSz: 324
    lineSz: 64
    sideSz: 64
    sideBwtSz: 48
    sideBwtLen: 192
    numSides: 7
    numLines: 7
    ebwtTotLen: 448
    ebwtTotSz: 448
    color: 0
    reverse: 0
Reading plen (2): 11:09:13
assert_gt: expected (0) > (0)
ds.h:665
bowtie2-inspect-s: ds.h:665: T& EList<T, S>::back() [with T = std::basic_string<char>; int S = 128]: Assertion `0' failed.
```
Trying to map anything on this index (using `bowtie2 -x test-corrupted`) results in:
``` bash
(ERR): bowtie2-align died with signal 11 (SEGV)
```

The problem disappears if the "gap-only" segment is not the first one in the fasta file.
``` fasta
>DUMMY_1
AAAGACTGCGCAATTTCTGAAAACTGAATTCGAATAAATTTAGGGTTGAAGTCATTTAAA
TTTCCTGTGAAGTTGTTGGTTCGTGTGTTGATAAGTGATGAGCTGATTGACTGATACGAA
TTGTTTTGACGCTTGCATCTTTAGACTTTTTTGTACGGGGTGTTGCAAGCCAGTTCTCAA
ACAATGCGTGCCAAGATCCAGCCAGTCAAAGCGATGTGACTCAGCGTCACGAGCACAGAT
GCACCGTGCAGCCATTTGAAACGCAGCCGGTCATCAGAGTCAGTGGCTCGATTGATTGCT
GGCATCAGCAACTGCCGCACGGGGATGGTCGTTAAAGCAACGGCCGCCATAACAATTGCA
GCCATGGTGTCGTGAGACCACCACAATAAAGCGGCAACGGCCGCCGTAGCCACCACAAAC
AAATAGAACCAAGGAAAAGCAAGCCGCAGCGTAGCACCAGCAGATGGGGCTGGAAGTGCT
TTGAATAAAAAAGCAGCAAAGCCAAAGGAGTACAGGGTCATGCCGCCGAACAACAGCGCG
GTGACCAATAT
>DUMMY_0
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNN
>DUMMY_2
TGGATAGGCAAAAAACAAAAAAGCCGCCTGGTGAACCAGGCGGCTTTGTTGGGGTGAATT
GACAGAACTTTTGTTCAAAGGCATAAAAACATTATGCACCAGCAACGTGACACCCTCGAA
CAGCGGTATCAAGCAGAAGTAGTGGCCTTTTTCAAATCCTTGACGGCTTGAAGAACTTCC
TCAACATGGCCTGGGACTTTGAGGCCTCTCCATTCGTTTTTGAGGGTGCCATCAGGGCCG
ACCAAAAATGTACTGCGCTCAATGCCTTTGACTTTCTTGCCGTACATGATTTTGTTTTTG
ACCACGCCAAACATGTGGCACATCTTTTCTTCCGTGTCCGCAATGAGTTCAAAGGGCAAC
TCAAGTTTGGCTTTGAAATCATCGTGGGACTTCATGTTATCGCGTGAGACACCCAAAACA
AGAGCACCAGCCTTGACGAAATCTTTGTATTTGTCGCGGAACTGCATGGCTTCTGTGGTG
CAACCCGGTGTGTTGTCTTTCGGGTAGAAAAACAAAACAACGGTTTGGCCGACGTGAGAG
ATGTTCGAAACTTTCAATCCGCTGGTGGCGTTGGCTTCAAATTCGGGGAGGGGTTTGTTC
AGAACGATCGCCATGGAACTTTAACTCGCTAGTCTGGTTGTGCGTACAGCACCTGCATGT
GGTGCCATTGCCGCTGGAGGCCAATAACCTCGGCAACCAGGTATTTTACTCTGAACGGCG
TTTTGACAGGCAACAACTGCCCA
```
Gives
```
$ bowtie2-build testseq.fna test-notcorrupted
$ bowtie2-inspect -v -n test-notcorrupted
Input ebwt file: "test-notcorrupted"
Output file: ""
Local endianness: little
Assertions: enabled
Trying test-notcorrupted
DUMMY_1
DUMMY_2
```

Additional information
----------------------------

- Operating system: Debian 10 (Buster, Testing)
- Bowtie2 version: 2.3.4.1
``` bash
$ bowtie2 --version
/home/testuser/.miniconda3/envs/testenv/bin/bowtie2-align-s version 2.3.4.1
64-bit
Built on prealloc-abbgtk4t-bbf9c480-d1fa-4afa-af46-fcb0649b95f5
Tue Feb 13 07:04:25 UTC 2018
Compiler: gcc version 4.8.5 (GCC) 
Options: -O3 -m64 -msse2 -funroll-loops -g3 -std=c++98 -DPOPCNT_CAPABILITY -DWITH_TBB -DNO_SPINLOCK -DWITH_QUEUELOCK=1
Sizeof {int, long, long long, void*, size_t, off_t}: {4, 8, 8, 8, 8, 8}
```
- Install method: conda

**Comment by nigiord** [comment]
Actually things are worse than expected: the index is ill-built after the first "gap-only" sequence. That means that bowtie2 will not map anything on sequences that follow a "gap-only" sequence in the index. In the second example above, it means that `bowtie2-align` will apparently work, but will only map reads on DUMMY_1 without displaying any error.

Furthermore, while `bowtie2-inspect -n` works in the second case (sequence names are correctly displayed), the default command `bowtie2-inspect` returns an error:
```bash
assert_eq: expected (123, 0x7b) got (122, 0x7a)
bt2_inspect.cpp:233
bowtie2-inspect-l: bt2_inspect.cpp:233: void print_ref_sequences(std::ostream&, bool, const EList<std::basic_string<char> >&, const TIndexOffU*, const string&): Assertion `0' failed.
```
Overall, despite the warnings, gap-only sequences are currently not handled by bowtie2. Worst, their presence at the index phase silently generate incoherent results at the mapping phase.

**Comment by ch4rr0** [comment]
I have a fix for this issue particular but unfortunately, it's breaking one of our other alignment modes. I am working to have this fixed in time for our upcoming release (slated for next week).

**Comment by ch4rr0** [comment]
This issue has been fixed in our most recent release, v2.3.4.2
